### ISSUE TITLE
<amp-list> renders content in wrong location

### ISSUE THREAD

**Issue by rowan-m** [ampproject/amphtml#4255]
## Short description of your issue:

Using `<amp-list>` to insert `<tr>` inside a `<table>` will actually result in the `<amp-list>` rendering before the `<table>` tag. It should either render within the table or provide an error.

``` html
    <table>
      <amp-list width=auto height=100 layout=fixed-height
                src="https://ampbyexample.com/json/examples.json">
        <template type="amp-mustache" id="amp-template-id">
          <tr><td>{{title}}</td></tr>
        </template>
      </amp-list>
    </table>
```
## How do we reproduce the issue?

https://merewood.org/amp-list-transform.html#development=1
## What browsers are affected?

Chrome, Samsung browser, probably all?
## Which AMP version is affected?

1469143772973

**Comment by ericlindley-g** [comment]
@dvoytenko passing to you for triage, as I think you've done the most work on amp-list

**Comment by dvoytenko** [comment]
/cc @Gregable 

This is an interesting case. Basically, as we operate on the constructed DOM, there's no way for `amp-list` to determine that it's inside the `<table>` tag as the element itself is outside of it. I think as a first step we need to recognize this fact by disallowing `amp-list` to be directly inside `<table>` and `<tr>` elements.

**Comment by Gregable** [comment]
We don't have a disallowed_parent mechanism at the moment, so I'm trying to figure out how to quickly test how much breakage this would cause.

How would this work in the given example? Presumably the tag would need to be inside the `<table>` in order to repeat the rows, no?

**Comment by Gregable** [comment]
I see, you would modify this to be something like:

```html
<amp-list width=auto height=100 layout=fixed-height
          src="https://ampbyexample.com/json/examples.json">
  <table>
    <template type="amp-mustache" id="amp-template-id">
      <tr><td>{{title}}</td></tr>
    </template>
  </table>
</amp-list>
```

In practical terms, the validator can detect this currently, but it's only able to do so because our html parser has a bug that doesn't cause it to serialize the tag stream the same way the browser does in this case.

I am a little hesitant to add a validation constraint that requires the validator to parse HTML incorrectly. We may later need to add a validation constraint that requires the validator to make this DOM transformation.

Said another way, the example doc and this doc:

```html
      <amp-list width=auto height=100 layout=fixed-height
                src="https://ampbyexample.com/json/examples.json">
        <template type="amp-mustache" id="amp-template-id">
          <tr><td>{{title}}</td></tr>
        </template>
      </amp-list>
    <table>
    </table>
```

Are identical according to the HTML parsing spec, but you are asking for one of them to be valid and the other invalid.

**Comment by dvoytenko** [comment]
Putting `<table>` inside `amp-list` kind of defies its purpose. Basically, currently `amp-list` doesn't work with `<table>`. We are considering a separate `<amp-render>` element that would just allow the author to do their own iteration inside which would allow all this.

And, yes, I totally see your point. This is really an annoying part of html parsing.

How about this instead: can we ban `<tr>` and `<td>` children in `<template>`?

**Comment by dvoytenko** [comment]
I mean direct children, not deep inside `<template>`.

**Comment by Gregable** [comment]
children, not descendants. That's definitely doable. I imagine we'd want the same constraint for `<tbody>`, `<thead>`, `<tfoot>`, `<th>`?

**Comment by dvoytenko** [comment]
Correct.

**Comment by Gregable** [comment]
This has been in the backlog for some time, so I took a look at implementing it. It requires some new code to the validator, nothing major, but it can't be implemented as a rules only change. 

I also tried to find some examples of pages in the wild that would actually fail. I sampled 10,000 AMP documents that also contained both `amp-list` and `template` tags, and didn't find a single one where the above table tags were direct children of the `template` tag.

I have a working implementation of the necessary change which just needs review, but since it accumulates a little extra code complexity and I couldn't find any examples in my sampling, I'm wanted to see if anyone has thoughts on if this should be just left out of the validator rules instead, and closed as working as intended. I don't have a strong opinion myself but would probably default to simply closing this if nobody else has a preference.
